### PR TITLE
lxd: update to 3.17

### PR DESCRIPTION
**Pull request by julio641742**

### CODE REVIEW

**Comment by Duncaen** [comment]
Would be better if new dependencies and the update are in the same PR, to make merging and easier and TravisCI pass.

**Comment by julio641742** [comment]
Ok

**Comment by julio641742** [comment]
Everything is up and ready for review :+1:

File: srcpkgs/libco/files/LICENSE [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Copyright 2019 Free Ekanayaka
```
Comment (CameronNemo) [inline srcpkgs/libco/files/LICENSE]: do not vendor licenses

File: srcpkgs/libco/template [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Template file for 'libco'
+pkgname=libco
+version=1.0.1
+revision=1
+short_desc="Cooperative multithreading library written in C89"
+maintainer="Julio Galvan <julio@epazote.net>"
+license="ISC"
```
Comment (CameronNemo) [inline srcpkgs/libco/template]: According to the headers of the files themselves, it is in public domain

File: srcpkgs/libco/template [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+homepage="https://github.com/freeekanayaka/libco"
+distfiles="https://github.com/freeekanayaka/libco/archive/v${version}.tar.gz"
+checksum=b5a49b0352e52edffc0046b592e4414945a2b637f7433f8c526b8cdaab9d24bd
+shlib_provides="libco.so"
```
Comment (CameronNemo) [inline srcpkgs/libco/template]: pretty sure that you can remove this

File: srcpkgs/raft/template [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+checksum=d2bc3c15db76a817802898ed0dba9d5608de12839a9d64b45ba3bd39c126eccf
+
+pre_configure() {
+	autoreconf -i
+}
+
+pre_build() {
+	case "$XBPS_TARGET_MACHINE" in
+		# fix musl missing error.h header
+		*-musl) sed -i '/error.h/d' src/replication.c;;
```
Comment (CameronNemo) [inline srcpkgs/raft/template]: please report upstream

File: srcpkgs/libco/template [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+do_build() {
+	sed -i 's/gcc/$(CC)/g' Makefile
```
Comment (CameronNemo) [inline srcpkgs/libco/template]: please send a patch upstream

**Comment by Anachron** [comment]
Please fix remaining issues so we can ship the latest lxd package.

File: srcpkgs/raft/template (hunk 4 above)
Comment (julio641742) [reply srcpkgs/raft/template]: reported (here)[https://github.com/CanonicalLtd/raft/issues/52]

File: srcpkgs/libco/template (hunk 5 above)
Comment (julio641742) [reply srcpkgs/libco/template]: [here](https://github.com/freeekanayaka/libco/pull/3)

**Comment by julio641742** [comment]
Ready for another review :+1:

**Comment by CameronNemo** [comment]
LGTM.

**Comment by CameronNemo** [comment]
Can you rebase on current master to fix the conflicts? This looks good from my end.

**Comment by julio641742** [comment]
Sorry for the late response @CameronNemo 
everything is up and ready for merge

**Comment by CameronNemo** [comment]
hmm lxd commit on this pr too should be fine.

File: srcpkgs/libco/template (hunk 3 above)
Comment (julio641742) [reply srcpkgs/libco/template]: without this line, building lxd gives trouble
```
...
=> ERROR: lxd-3.15_1: failed to install 'lxc-devel-3.1.0_1' dependency! (error 8)
dqlite-0.9.1_1: broken, unresolvable shlib `libco.so'
Transaction aborted due to unresolved shlibs.
=> ERROR: Please see above for the real error, exiting...
```

**Comment by julio641742** [comment]
lxd commit was added and is ready to be merged

**Comment by CameronNemo** [comment]
duncaen requested that libco use the gnu-makefile build style and the do_build() function be removed.

**Comment by julio641742** [comment]
@CameronNemo done

**Comment by q66** [comment]
While at it, you can import this patch for `libco` https://gist.github.com/q66/7fa82b0f4377dbdcee1b82972e7b7d58

This adds assembly support for aarch64 and ELFv2 based ppc64 and reenables the old assembly code for 32-bit ppc (and ELFv1 ppc64, but that's irrelevant to us). It's taken right from `libco`'s upstream (higan), the repo you're linking to is technically a downstream.

**Comment by julio641742** [comment]
@q66 patch applied

File: srcpkgs/libco/template (hunk 3 above)
Comment (the-maldridge) [reply srcpkgs/libco/template]: You should not need this line given that you have added libco to the common/shlibs file.  Please remove and re-test.

File: srcpkgs/libco/template [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+short_desc="Cooperative multithreading library written in C89"
+maintainer="Julio Galvan <juliogalvan@protonmail.com>"
+license="Public Domain"
+homepage="https://github.com/freeekanayaka/libco"
+distfiles="https://github.com/freeekanayaka/libco/archive/v${version}.tar.gz"
+checksum=8876aa8e61f07b37adb3ad422b369ecf9bf6fad20ce691428c35719ba5f16615
+shlib_provides="libco.so"
+
+do_install() {
+	vinstall libco.so 0664 /usr/lib/ libco.so.${version}
```
Comment (the-maldridge) [inline srcpkgs/libco/template]: This looks an awful lot like you're taking sonames of variable versions and linking them to a fixed .1 version.  This is almost certainly wrong.

File: srcpkgs/libco/template (hunk 3 above)
Comment (CameronNemo) [reply srcpkgs/libco/template]: I have just tested with the line removed. libco package is built without the shlibs provides property set correctly, it does not provide libco.so without the line.

Comment (q66) [reply srcpkgs/libco/template]: The thing is, it should provide the versioned one, not `libco.so` per se...

Comment (CameronNemo) [reply srcpkgs/libco/template]: I did not notice a versioned one in the github repo. Does upstream libco provide a versioned library?

**Comment by CameronNemo** [comment]
There has been a new release of `lxd`, 3.16. Also sqlite-replication had a new release.

File: srcpkgs/libco/template (hunk 3 above)
Comment (q66) [reply srcpkgs/libco/template]: It does not. But it should - if it does not there is no point in versioning it manually in the template (the linkage will always be wrong). I reported this in the upstream repo

Comment (q66) [reply srcpkgs/libco/template]: This is why:

```
q66@moca: /home/q66$ gcc -fPIC foo.c -shared -o libfoo.so.1
q66@moca: /home/q66$ ln -s /home/q66/libfoo.so.1 libfoo.so
q66@moca: /home/q66$ gcc test.c -o test -L. -lfoo
q66@moca: /home/q66$ ldd test
        linux-vdso64.so.1 (0x00003fa1975d0000)
        libfoo.so => not found
        libc.so.6 => /usr/lib/libc.so.6 (0x00003fa1973a9000)
        /lib/ld64.so.2 (0x00003fa1975d8000)
q66@moca: /home/q66$ gcc -fPIC foo.c -shared -o libfoo.so.1 -Wl,-soname,libfoo.so.1
q66@moca: /home/q66$ gcc test.c -o test -L. -lfoo
q66@moca: /home/q66$ ldd test
        linux-vdso64.so.1 (0x00003f7af9820000)
        libfoo.so.1 => not found
        libc.so.6 => /usr/lib/libc.so.6 (0x00003f7af95f9000)
        /lib/ld64.so.2 (0x00003f7af9833000)
```

Comment (q66) [reply srcpkgs/libco/template]: If you version it manually just by creating symlinks this also means that once you don't have the devel package installed the dynamic linker will fail to find the runtime library afterwards. So at very least you would also need to provide the appropriate linker flag and then remove the `shlib_provides`.

**Comment by CameronNemo** [comment]
Please apply the patch from https://github.com/canonical/dqlite/pull/161

**Comment by CameronNemo** [comment]
How is this branch working? Are you able to successfully run a container? I had trouble with 3.18 from #15045.

**Comment by Anachron** [comment]
LXD is on 3.20 now, please someone close.

**Pull request closed without merging**